src/GitVersionCore/GitVersionCacheKeyFactory.cs

```diff
@@ -23,10 +25,104 @@ private static string GetGitSystemHash(GitPreparer gitPreparer, IFileSystem file
+            var traverse = TraverseTree(Path.Combine(dotGitDirectory, "refs"));
 
-            return GetHash(dotGitDirectory, lastGitRefsChangedTicks.ToString());
+            return GetHash(traverse.ToArray());
+        }
+
+        // lifted from https://msdn.microsoft.com/en-us/library/bb513869.aspx
+        public static List<string> TraverseTree(string root)
+        {
+            var result = new List<string> { root };
+
+            // Data structure to hold names of subfolders to be
+            // examined for files.
+            var dirs = new Stack<string>(20);
+
+            if (!Directory.Exists(root))
+            {
+                throw new ArgumentException();
+            }
+
+            dirs.Push(root);
+
+            while (dirs.Any())
+            {
+                string currentDir = dirs.Pop();
+
+                var di = new DirectoryInfo(currentDir);
+                result.Add(di.Name);
+
+                string[] subDirs;
+                try
+                {
+                    subDirs = Directory.GetDirectories(currentDir);
+                }
+                // An UnauthorizedAccessException exception will be thrown if we do not have
+                // discovery permission on a folder or file. It may or may not be acceptable 
+                // to ignore the exception and continue enumerating the remaining files and 
+                // folders. It is also possible (but unlikely) that a DirectoryNotFound exception 
+                // will be raised. This will happen if currentDir has been deleted by
+                // another application or thread after our call to Directory.Exists. The 
+                // choice of which exceptions to catch depends entirely on the specific task 
+                // you are intending to perform and also on how much you know with certainty 
+                // about the systems on which this code will run.
+                catch (UnauthorizedAccessException e)
+                {
+                    Logger.WriteError(e.Message);
+                    continue;
+                }
+                catch (DirectoryNotFoundException e)
+                {
+                    Logger.WriteError(e.Message);
+                    continue;
+                }
+
+                string[] files = null;
+                try
+                {
+                    files = Directory.GetFiles(currentDir);
+                }
+
+                catch (UnauthorizedAccessException e)
+                {
+                    Logger.WriteError(e.Message);
+                    continue;
+                }
+
+                catch (DirectoryNotFoundException e)
+                {
+                    Logger.WriteError(e.Message);
+                    continue;
+                }
+
+                foreach (string file in files)
+                {
+                    try
+                    {
+                        var fi = new FileInfo(file);
+                        result.Add(fi.Name);
```

Comment:
file = C:\Users\tim\AppData\Local\Temp\2\TestRepositories\cc7f7d6e-5b1f-4f63-a760-960409c862f9\.git\refs\heads\master
fi.Name = master

Are you sure you want to repeat the path for every file? I was trying to reduce the amount of text to hash. Currently it goes from

![2016-07-25_0826](https://cloud.githubusercontent.com/assets/157766/17107072/908ef960-5242-11e6-88d3-267ea1ce35a6.png)

to

![2016-07-25_0827](https://cloud.githubusercontent.com/assets/157766/17107076/95fbe494-5242-11e6-8d39-b6760bf7f9a2.png)

I left the path in the first item of the results, but even that is not necessary because the hash computation should not care about the physical location of the repository. It should start at "refs". Maybe I'll make that change. What do you think?